RationaleMCP/0027/HindleyMilner.md

```diff
@@ -0,0 +1,336 @@
+
+#### The algorithm
+
+The Hindley-Milner algorithm modified to perform unit inference in Modelica:
+* Traverse the flattened equation system and collect constraints before any evaluation has occured.
+* After evaluable parameters have been evaluated, and expressions have been reduced to values when possible, evaluate the constraints, as well as their conditions. (This propagates `empty` and `undefined` units to the top of the constraints.)
+* Discard any constraints of the form _`u` must be equivalent to empty unit_, or _`u` must be equivalent to undefined unit_, as they are trivially satisfied.
+* Discard any constraints having at least a condition that was evaluated to `false`.
+* Represent all constraints in their single-sided form.
+* Start with an empty set `S` of substitutions.
+* Loop:
+    * Select a solvable constraint or one that can be checked, if no such constraint exists exit the loop.
+    * If the constaint is in the form of just a `well-formed` unit, check that it is equivalent to `"1"`, it is an error otherwise.
+    * If it can be solved for `var.unit`:
+        * Replace the constraint by the substitution _`var.unit` => `u`_, where `var.unit` is not present in `u`.
+        * Apply the substitution to the right-hand side of the substitutions already in `S`.
+        * Add the substitution to `S`.
+        * Apply the substitution to all remaining constraints.
+* The unit of a variable for which there is no substitution in `S` is unknown.
+* For a variable `var` that has a substitution `var.unit` => `u` in `S`:
+    * If `u` is a `well-formed` unit, then this is the unit of `var`. (Given that no unit errors have been detected, a variable with a `unit`-attribute will get the given unit.)
+    * Otherwise, the unit of `var` is unknown.
+
+## Unit constraints of a Modelica model
+
+The following rules describe the unit meta-expression of a given Modelica expression, where it makes sense.
+They also provide the constraints that arise from different Modelica constructs.
+
+When it is specified that _`u1` must be equivalent to `u2`_, the corresponding constraint should be collected.
+
+In this section, the unit meta-expression of expression `e` is called `e.unit` and conditions of conditional constraints are highlighted with [].
+
+### Variables
+If `var` has a declared `unit`-attribute, `var.unit` must be equivalent to it.
+
+`var.unit` must be equivalent to the unit of the following attributes if present:
+ * `start`
+ * `min`
+ * `max`
+ * `nominal`
+
+(`var.unit` should be convertible to the `displayUnit`-attribute of `var` if it is present, but there is no corresponding constraints for the unit inference.)
+
+Consider the expression `e` of the form `var`, where `var` is a Modelica variable.
+* If `var` doesn't have a declared unit attribute, and has constant component variablity, `e.unit` is the `empty` unit.
```

Comment:
I believe this is the refined rule for constants which we agreed to at the meeting in Linköping:
```suggestion
* If `var` is a constant without declared `unit`-attribute, `e.unit` is determined by evaluating the unit of the declaration equation.
  When the declaration equation's unit is well-formed, it is propagated to `e.unit`.
  Otherwise, `e.unit` is the `empty` unit, preventing it from being determined by general unit inference.
```